Update Onboarding project -->

subh2310 opened the pull request.

subh2310 committed:
Initial Commit after make changes

subh2310 committed:
Updated Readme File

subh2310 committed:
Modification on Readme File

abhi-1408 commented:
Please add desc in the PR mentioning the changes it contains.

Also going fwd please try to follow [this](https://github.com/ssinha2103/UP_onboarding_project/pull/2#issuecomment-1291812176)

abhi-1408 commented on foodhub/accounts/models.py:
```diff
@@ -0,0 +1,145 @@
+        max_length=150, blank=True, null=True, unique=True)
+    USERNAME_FIELD = 'email'
+    is_active = models.BooleanField(default=True)
+    REQUIRED_FIELDS = []
```
why is `email` not in the `REQUIRED_FIELDS`

subh2310 replied:
As per requirement i have updated my models.py file.

abhi-1408 commented on foodhub/accounts/models.py:
```diff
@@ -0,0 +1,145 @@
+class OrderStatus(models.TextChoices):
+    ACTIVE = 'ACT', 'Active'
+    CANCELLED = 'CAN', 'Cancelled'
+    FINISHED = 'FIN', 'Finished'
```
Just a sugg, instead of Active it can be Placed, similarly instead of finished it can be Completed.

subh2310 replied:
As per requirement i have updated my models.py file

abhi-1408 commented on foodhub/accounts/models.py:
```diff
@@ -0,0 +1,145 @@
+    email = models.EmailField("Email", max_length=254, blank=False)
+    phone = models.CharField("Phone", max_length=20)
+    created_at = models.DateTimeField(
+        "Created At", auto_now_add=True, null=True)
```
`auto_now_add` is set to True and null is set to True as well, can u also explain as when will the situation happen.

subh2310 replied:
As per requirement i have updated my models.py file

abhi-1408 commented on foodhub/accounts/models.py:
```diff
@@ -0,0 +1,145 @@
+        "Created At", auto_now_add=True, null=True)
+    created_by = models.ForeignKey(
+        'User', on_delete=models.CASCADE, related_name='merchant_created_by', null=True)
+    updated_by = models.ForeignKey(
```
In which usecase is this needed ?

abhi-1408 replied:
Also the `related_name` can be named better.

subh2310 replied:
As per requirement i have updated my models.py file

abhi-1408 commented on foodhub/accounts/models.py:
```diff
@@ -0,0 +1,145 @@
+class User(AbstractUser):
+    username = None
+    email = models.EmailField(
+        max_length=150, blank=True, null=True, unique=True)
```
Also why email is set with blank=True, null=True ?

subh2310 replied:
As per requirement i have updated my models.py file

abhi-1408 commented on foodhub/accounts/models.py:
```diff
@@ -0,0 +1,145 @@
+    lon = models.FloatField("Longitude", blank=True, null=True)
+    lat = models.FloatField("Latitude", blank=True, null=True)
+    active = models.BooleanField("Active", default=False)
+    items = models.ManyToManyField(Item)
```
Please add a `related_name` to items as well.

subh2310 replied:
As per requirement i have updated my models.py file

abhi-1408 commented on foodhub/accounts/models.py:
```diff
@@ -0,0 +1,145 @@
+    Optional Fields : Description
+    """
+    name = models.CharField("Name", max_length=200)
+    merchant = models.ForeignKey(Merchant, on_delete=models.CASCADE)
```
Please add a related name to `merchant` as well.
Also add `related_name` where ever needed in `Order` & `Store` models

subh2310 replied:
As per requirement i have updated my models.py file

abhi-1408 commented on foodhub/accounts/models.py:
```diff
@@ -0,0 +1,145 @@
+                              max_length=100,
+                              choices=OrderStatus.choices,
+                              default=OrderStatus.ACTIVE)
+    items = models.ManyToManyField(Item)
```
How will it be created in db, a `items` m2m field in order table

subh2310 replied:
As per requirement i have updated my models.py file

abhi-1408 commented on foodhub/accounts/serializers.py:
```diff
@@ -0,0 +1,113 @@
+
+
+class RegisterSerializer(serializers.ModelSerializer):
+    email = serializers.EmailField(required=True, validators=[
```
Instead of this, u can add `unique=True` on email field itself.

subh2310 replied:
As per requirement i have updated my serializers.py file

abhi-1408 commented on foodhub/accounts/serializers.py:
```diff
@@ -0,0 +1,113 @@
+        Perform object level validation for Merchant Integrity during Store
+        Creation.
+
+        validate_items_merchant : Checks if the Items' Merchant is same as
```
why do we need to check during store creation that Items' Merchant is same as Order's merchant ?

subh2310 replied:
As per requirement i have updated my serializers.py file

abhi-1408 commented on foodhub/accounts/serializers.py:
```diff
@@ -0,0 +1,113 @@
+        Perform object level validation for Merchant Integrity during Order
+        Creation.
+
+        Contains below functions :-
```
I hope here's what u wanted to do, validating that the respective item is from the store mentioned.
Can u optimise ur approach here.

subh2310 replied:
As per requirement i have updated my serializers.py file

abhi-1408 commented on foodhub/accounts/tasks.py:
```diff
@@ -0,0 +1,30 @@
+log = structlog.get_logger()
+
+
+@shared_task(name="create_order", autoretry_for=(OperationalError,), retry_kwargs={
```
what adv it served to have it as an async task

subh2310 replied:
As per requirement i have updated my tasks.py file

abhi-1408 commented on foodhub/accounts/views.py:
```diff
@@ -0,0 +1,239 @@
+
+log = structlog.get_logger()
+
+class RegisterView(APIView):
```
why did u preferred class based views instead of function based views.

abhi-1408 commented on foodhub/accounts/views.py:
```diff
@@ -0,0 +1,239 @@
+
+log = structlog.get_logger()
+
+class RegisterView(APIView):
```
why did u preferred class based view instead of function based view.

subh2310 replied:
As per requirement i have updated my views.py file

abhi-1408 commented on foodhub/accounts/views.py:
```diff
@@ -0,0 +1,239 @@
+        The paginator instance associated with the view, or `None`.
+        """
+        user_agent = self.request.headers.get('User-Agent', None)
+        log.info('INSIDE paginator',user_agent = user_agent)
```
just a sugg, try to add events as snake_case.

subh2310 replied:
As per requirement i have updated my views.py file

abhi-1408 commented on foodhub/foodhub/settings.py:
```diff
@@ -0,0 +1,166 @@
+# See https://docs.djangoproject.com/en/4.1/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = 'django-insecure-j#gy^4ppz59i&w029%gf#*6nn4e0&ovbt6j%ic6kjj&z2yx#=u'
```
Try to always have secrets stored & loaded from ENV

subh2310 replied:
Sure i will create new .env file and separate Secret Key, Database Configuration and rest of the things

abhi-1408 commented on foodhub/foodhub/settings.py:
```diff
@@ -0,0 +1,166 @@
+    'rest_framework',
+    'django_filters',
+    'accounts',
+    'silk',
```
also try to compare how does the locust stats changes when u have silk enabled vs not.

subh2310 replied:
Ok sure i will do it

abhi-1408 commented on foodhub/foodhub/settings.py:
```diff
@@ -0,0 +1,166 @@
+        'HOST': '127.0.0.1',
+        'PORT': '3306',
+        }
+    }
```
Similarly here also, try to use it from ENV

subh2310 replied:
Sure i will do it.

abhi-1408 commented on foodhub/tests/api/test_user.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@pytest.mark.django_db
+def test_merchant_creation_api_endpoint(user):
```
Also check for the creation of the instance in db.

subh2310 replied:
As per requirement i have created new tests folder and define two file there conftest.py and tests.py and written a test cases

abhi-1408 commented:
As discussed on call @subh2310,
Needed a few more changes to fine tune it:
- register, login url should be shown as well
- changes in user registration & login form 
- user can be  like merchant, customer which should also be present
- all changes should be done through UI as well. 


Lemme know once the changes are done, will continue our discussion.

abhi-1408 commented:
@subh2310 As discussed,
- users should be fixed
- items logic should be fixed
- should be able to place orders
- The get response for items should show the store name instead of store id.

Will continue the discussion tmrw.

subh2310 committed:
Updated Model in models.py file

As discussed I have created new model name Profile and define merchant and consumer on it.

subh2310 committed:
No need of this file for validation.

subh2310 committed:
Updated serializers.py file

I have added Profile Serializer for Merchant and consumer and Created ViewConsumerSerializer for validating the Consumer .

subh2310 committed:
Created permissions.py file 

In this file I have created Merchant Permission and Consumer Permission which is going to define what permission merchant and consumer have it.

subh2310 committed:
Updated tasks.py file

Added one more task on it that is create_store and updated create_orders.

subh2310 committed:
Updated urls.py file 

Added urls like : place_order, see_order, view-consumers

subh2310 committed:
Updated views.py file

Added more view on it like : SeeOrderView, ViewConsumerView, placeOrderView and UserChanged PasswordView and updated on items and Store view.

subh2310 committed:
Added locustfile.py

Added locustfile.py for load testing

subh2310 committed:
Updated admin.py file.

subh2310 committed:
Updated confest.py file

Removed all code and added two fixture on it : merchant_data and consumer_data.

subh2310 committed:
Updated tests.py file

subh2310 committed:
Updated requirement.txt file.

As per the requirement related to my Onboarding project i have installed so many module and added in this file

subh2310 committed:
Updated settings.py file

As per the requirement of the project  I have updated settings.py file

subh2310 committed:
Updated urls.py file

subh2310 committed:
Update README file

subh2310 commented:
I have updated my onboarding Project and everything is working.
